### PR TITLE
Trainer for a fixed number of epochs

### PR DESCRIPTION
**Pull request by W4ngatang**
Includes:

* train for a maximum number of epochs (passes over all of a task's data)

* refactoring of stop check in trainer

* removal of a possible infinite loop bug

### CODE REVIEW

File: src/trainer.py [hunk 1]
```diff
@@ -312,7 +317,11 @@ def train(self, tasks, stop_metric,
         task_infos, metric_infos = self._setup_training(tasks, batch_size, train_params,
                                                         optimizer_params, scheduler_params, phase)
 
-        if shared_optimizer:  # if shared_optimizer, ignore task_specific optimizers
+        if shared_optimizer:  # If shared_optimizer, ignore task_specific optimizers
+            if self._max_epochs_per_task:
+                # TODO(Alex or Yada): make this generalizable to not depend
```
Comment (W4ngatang) [inline src/trainer.py]: what does this mean?

**Comment by W4ngatang** [comment]
Ready for review, though I need to run a test.

File: src/trainer.py (hunk 1 above)
Comment (pruksmhc) [reply src/trainer.py]: Actually revert this, I think this was from when I was trying to figure out how _max_epochs_per_task worked

File: config/defaults.conf [hunk 2]
```diff
@@ -123,6 +123,7 @@ val_data_limit = 5000  // Maximum number of examples to be used during mid-train
                        // end of main.py that is invoked by do_full_eval.
 val_interval = 1000  // Number of steps between validation checks.
                      // Must be divisible by bpp_base (which is usually 1).
+max_epochs_per_task = None // If positive, maximum number of epochs (full pass over all training data) to perform
```
Comment (sleepinyourhat) [inline config/defaults.conf]: Explain how this relates to max_vals. We run until _either_ criterion is hit, right? Might also be worth adding an option to set max_vals to None, with the same effect.

Comment (sleepinyourhat) [reply config/defaults.conf]: Also, why 'per task'?

File: src/trainer.py [hunk 3]
```diff
@@ -312,7 +317,11 @@ def train(self, tasks, stop_metric,
...
+                # on max_epochs_per_tasks.
+                n_epoch_steps = sum([info["n_tr_batches"] for info in task_infos.values()])
```
Comment (sleepinyourhat) [inline src/trainer.py]: This is never used, delete.

File: src/trainer.py [hunk 4]
```diff
@@ -699,6 +708,15 @@ def _check_stop(self, epoch, stop_metric, tasks):
         ''' Check to see if should stop '''
         task_infos, metric_infos = self._task_infos, self._metric_infos
         g_optimizer = self._g_optimizer
+
+        if self._max_epochs_per_task > 0:
+            for task in tasks:
+                task_info = task_infos[task.name]
+                n_epochs_trained = task_info['total_batches_trained'] / task_info['n_tr_batches']
+                if n_epochs_trained >= self._max_epochs_per_task:
+                    log.info("Maximum batches trained on %s.", task.name)
```
Comment (sleepinyourhat) [inline src/trainer.py]: => 'epochs'

**Pull request merged**